### PR TITLE
fix: network verification for collision network + local network

### LINKED ISSUE #23242: [Bug]: Potential scam alert warning for Wethio network
### Describe the bug

Hi team, we have our own blockchain - Wethio Blockchain - running from June 2020, following are the details about our blockchain network -
Docs - https://docs.wethio.io
Wallet - https://wallet.wethio.io
Explorer - https://explorer.wethio.io

Our network RPC details are as follows
chain id - 78
rpc url - https://rpc.wethio.io
currency symbol - ZYN

In the latest release of metamask we are facing an issue where in our network is getting marked as Spam (pic attached below), because the chain id 78 is linked with the PrimusMoney blockchain on https://chainid.network/.

As asked by the technical support team, we created an issue on the Chains repository, but they have closed the issue saying that they cannot allow collisions.
Issue link - https://github.com/ethereum-lists/chains/issues/4423

But, when we started working on our blockchain in 2020 we checked for other chains with chain id 78 and did not find any. If Primus Money was listed on the Chains repository at that time, then it would have been our fault for using the chain id 78. But Primus money added their network info here in March 2021. So our only fault here is that not adding our network details on the Chains repository before Primus Money.

So we ask you to please look into this, so that at least the Potential Scam warning can be removed for our network.

### Expected behavior

We were not getting the scam warnings in earlier versions of Metamask

### Screenshots/Recordings

Here is the screenshot of the warning being displayed
Skype_Picture_2024_01

### Steps to reproduce

1. Go to the networks tab
2. Add Wethio network with the rpc details mentioned above
3. Potential Scam warning when trying to interact with the network

### Error messages or log output

_No response_

### Version

11.7.4

### Build type

None

### Browser

Firefox

### Operating system

Linux

### Hardware wallet

_No response_

### Additional context

_No response_

### Severity

_No response_

### PR DESCRIPTION
**Pull request by salimtb**
## **Description**

 here are the different fixes that this PR contains:

collision of two networks with the same chainId but with two different symbols ==> solution: manage this as a special case, add the other network and extend the list https://chainid.network/chains.json

scam warning displayed for networks running on localhost ==> exclude from network verification all networks running on localhost.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/23802?quickstart=1)

## **Related issues**

Fixes: #23242 

## **Manual testing steps**

1. Go to add network page
2. Go to custom add network form
3. Add the network with chainID 78 and rpc https://rpc.wethio.io/ and symbol ZYN

### **Before**

Screenshot 2024-03-29 at 10 49 00

Screenshot 2024-03-29 at 10 49 18

### **After**
Screenshot 2024-03-29 at 10 35 41

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by wethioproject** [comment]
any update ?

**Comment by salimtb** [comment]
> any update ?

@wethioproject the PR is under review process and should be merged soon

**Comment by wethioproject** [comment]
thanks for update

**Review by DDDDDanica — DISMISSED** [review]

**Comment by DDDDDanica** [comment]
LGTM !Thanks for the fix !

File: ui/hooks/useIsOriginalNativeTokenSymbol.js [hunk 1]
```diff
@@ -18,12 +37,28 @@ export function useIsOriginalNativeTokenSymbol(chainId, ticker, type) {
...
         if (mappedCurrencySymbol) {
           setIsOriginalNativeSymbol(mappedCurrencySymbol === ticker);
           return;
         }
 
+        const mappedAsNetworkCollision =
+          CHAIN_ID_TO_CURRENCY_SYMBOL_MAP_NETWORK_COLLISION[chainId];
+
+        if (mappedAsNetworkCollision) {
+          if (mappedAsNetworkCollision === ticker) {
```
Comment (darkwing) [inline ui/hooks/useIsOriginalNativeTokenSymbol.js:56]: Can these two `if`s be simplified?

File: ui/hooks/useIsOriginalNativeTokenSymbol.js [hunk 2]
```diff
@@ -1,15 +1,34 @@
...
+  ticker,
+  type,
+  rpcUrl = null,
+) {
   const [isOriginalNativeSymbol, setIsOriginalNativeSymbol] = useState(null);
   const useSafeChainsListValidation = useSelector(
     useSafeChainsListValidationSelector,
   );
+
+  const isLocalhost = (urlString) => {
```
Comment (darkwing) [inline ui/hooks/useIsOriginalNativeTokenSymbol.js:23]: Could we use the existing `isLocalhostOrHttps` here?

File: ui/pages/settings/networks-tab/networks-form/networks-form.js [hunk 3]
```diff
@@ -109,7 +110,7 @@ const NetworksForm = ({
   const [rpcUrl, setRpcUrl] = useState(selectedNetwork?.rpcUrl || '');
   const [chainId, setChainId] = useState(selectedNetwork?.chainId || '');
   const [ticker, setTicker] = useState(selectedNetwork?.ticker || '');
-  const [suggestedTicker, setSuggestedTicker] = useState('');
+  const [suggestedTicker, setSuggestedTicker] = useState([]);
```
Comment (darkwing) [inline ui/pages/settings/networks-tab/networks-form/networks-form.js:113]: Should we create a constant called `DEFAULT_SUGGESTED_TICKER` and set it to `[]` so we aren't needing to set it to `[]` in every place it gets reset?

File: ui/pages/settings/networks-tab/networks-form/networks-form.js [hunk 4]
```diff
@@ -446,19 +465,27 @@ const NetworksForm = ({
           (chain) => chain.chainId.toString() === decimalChainId,
         );
 
+        const matchedSymbols = safeChainsList.current?.reduce(
```
Comment (darkwing) [inline ui/pages/settings/networks-tab/networks-form/networks-form.js:468]: It looks like this code/logic gets repeated -- can we create a helper function for this?

**Review by darkwing — CHANGES_REQUESTED** [review]

File: ui/pages/settings/networks-tab/networks-form/networks-form.js (hunk 3 above)
Comment (chris6090) [reply ui/pages/settings/networks-tab/networks-form/networks-form.js]: kriseng83@gmail.com

File: ui/hooks/useIsOriginalNativeTokenSymbol.js (hunk 1 above)
Comment (chris6090) [reply ui/hooks/useIsOriginalNativeTokenSymbol.js]: kriseng83@gmail.com

File: ui/pages/settings/networks-tab/networks-form/networks-form.js (hunk 3 above)
Comment (chris6090) [reply ui/pages/settings/networks-tab/networks-form/networks-form.js]: > Should we create a constant called `DEFAULT_SUGGESTED_TICKER` and set it to `[]` so we aren't needing to set it to `[]` in every place it gets reset?

kriseng83@gmail.com

File: ui/hooks/useIsOriginalNativeTokenSymbol.js (hunk 1 above)
Comment (chris6090) [reply ui/hooks/useIsOriginalNativeTokenSymbol.js]: > Can these two `if`s be simplified?

kriseng83@gmail.com

Comment (chris6090) [reply ui/hooks/useIsOriginalNativeTokenSymbol.js]: > Can these two `if`s be simplified?

kriseng83@gmail.com

File: ui/pages/settings/networks-tab/networks-form/networks-form.js (hunk 3 above)
Comment (chris6090) [reply ui/pages/settings/networks-tab/networks-form/networks-form.js]: > Should we create a constant called `DEFAULT_SUGGESTED_TICKER` and set it to `[]` so we aren't needing to set it to `[]` in every place it gets reset?

Yes

File: ui/hooks/useIsOriginalNativeTokenSymbol.js (hunk 2 above)
Comment (chris6090) [reply ui/hooks/useIsOriginalNativeTokenSymbol.js:23]: Yes

File: ui/pages/settings/networks-tab/networks-form/networks-form.js (hunk 3 above)
Comment (chris6090) [reply ui/pages/settings/networks-tab/networks-form/networks-form.js]: > Should we create a constant called `DEFAULT_SUGGESTED_TICKER` and set it to `[]` so we aren't needing to set it to `[]` in every place it gets reset?

Yes

Comment (chris6090) [reply ui/pages/settings/networks-tab/networks-form/networks-form.js]: > Should we create a constant called `DEFAULT_SUGGESTED_TICKER` and set it to `[]` so we aren't needing to set it to `[]` in every place it gets reset?

File: shared/constants/network.ts [hunk 5]
```diff
@@ -643,6 +650,13 @@ export const CHAIN_ID_TO_CURRENCY_SYMBOL_MAP = {
     CHAINLIST_CURRENCY_SYMBOLS_MAP.ACALA_NETWORK,
 } as const;
 
+export const CHAIN_ID_TO_CURRENCY_SYMBOL_MAP_NETWORK_COLLISION = {
```
Comment (bergeron) [inline shared/constants/network.ts:653]: Would be good to add a comment above this describing its purpose, so people know when to add to it in the future

File: ui/hooks/useIsOriginalNativeTokenSymbol.js [hunk 6]
```diff
@@ -18,12 +37,26 @@ export function useIsOriginalNativeTokenSymbol(chainId, ticker, type) {
...
+        }
+
         const mappedCurrencySymbol = CHAIN_ID_TO_CURRENCY_SYMBOL_MAP[chainId];
         if (mappedCurrencySymbol) {
           setIsOriginalNativeSymbol(mappedCurrencySymbol === ticker);
           return;
         }
 
+        const mappedAsNetworkCollision =
+          CHAIN_ID_TO_CURRENCY_SYMBOL_MAP_NETWORK_COLLISION[chainId];
```
Comment (bergeron) [inline ui/hooks/useIsOriginalNativeTokenSymbol.js:53]: Would this approach work if there were multiple networks colliding with the same chain id?

**Comment by wethioproject** [comment]
Any eta to resolve issue ?

**Comment by sahar-fehri** [comment]
LGTM!

**Review by sahar-fehri — DISMISSED** [review]

**Comment by darkwing** [comment]
Could we write an E2E test here to automate this?  One basic network add and one collision case?  This would make me much more comfortable.

**Review by sahar-fehri — APPROVED** [review]

**Review by darkwing — APPROVED** [review]

**Pull request merged**